crates/bevy_utils/src/cow_arc.rs

```diff
@@ -0,0 +1,130 @@
+use std::{
+    fmt::{Debug, Display},
+    hash::Hash,
+    ops::Deref,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+/// Much like a [`Cow`](std::borrow::Cow), but owned values are Arc-ed to make clones cheap. This should be used for values that
+/// are cloned for use across threads and change rarely (if ever).
+pub enum CowArc<'a, T: ?Sized + 'static> {
+    /// A borrowed value
+    Borrowed(&'a T),
+    /// A static value reference
+    Static(&'static T),
+    /// An owned [`Arc`]-ed value
+    Owned(Arc<T>),
+}
+
+impl<'a, T: ?Sized> Deref for CowArc<'a, T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &Self::Target {
+        match self {
+            CowArc::Borrowed(v) => v,
+            CowArc::Static(v) => v,
+            CowArc::Owned(v) => v,
+        }
+    }
+}
+
+impl<'a, T: ?Sized> CowArc<'a, T>
+where
+    &'a T: Into<Arc<T>>,
+{
+    /// Converts this into an "owned" value. If internally a value is borrowed, it will be cloned into an "owned [`Arc`]".
+    /// If it is already an "owned [`Arc`]", it will remain unchanged.
+    #[inline]
```

Comment:
Just for my learning, what makes this kind of method desirable as `#[inline]` over methods like `clone`?

Reply:
It will turn a `CowArc<'a, T>` into a `CowArc<'static, T>` by creating an `Owned`/arc-ed `T` when necessary (ex: for `Borrowed` variants). Clone, on the other hand, will not created `Owned` values for `Borrowed` because it can just copy the reference.

Reply:
Oh wait i misread. You weren't asking about that :laughing:

Reply:
The others should probably be inlined too.